keps/sig-node/20200303-shared-pod-limits.md

```diff
@@ -0,0 +1,375 @@
+
+This proposal suggests a middle ground, and suggests a way to make it possible to describe to Kubernetes how to limit the resource consumption of multiple containers in the pod at once, on the Pod level, instead of trying to micro-manage the resource limits on the container level. For workloads where the work performed is burstable, this proposal would make it easier to allow the low-level mechanisms available in the underlying operating-system to manage the resources required for the task.
+
+### Goals
+
+This proposal aims to:
+
+*  Allow a `Burstable` pod to define that memory and CPU resource limits should be set on the level of the Pod.
+*  Prevent the developer from having to micro-manage the memory and CPU resource assignments for different containers in the same pod.
+*  Keep the current `Burstable` behavior as the default.  
+
+For workloads in the `Burstable` QoS level, the resource usage profile can be relativly dynamic. Memory and CPU can be required suddenly for a short period of time, and afterwards relinquished. By utilizing the Linux cgroup controllers this usage profile can be managed directly by the Linux kernel.
+
+### Non-Goals
+
+*  Providing a general-purpose interface to the full range of possible resource management provided by the Linux cgroup hierarchy.
+
+## Proposal
+
+The Pod QoS enhancement already implemented in Kubernetes manages resources as a hieararchy of cgroups in the following way:
+
+<pre>
+  QoS CGroup (one of guaranteed, burstable, or besteffort)
+   |
+   \ pod 
+      |
+      +-- container0 (pause container)
+      |
+      +-- container1 (first container)
+      |
+   = ... =
+      |
+      +-- containerN (N-th container)
+</pre>
+
+Each container level cgroup is limited based on the information provided in the `Container` specification. Currently, if the pod belongs to 
+the `Burstable` QoS level, and **all** of the containers specify limits for each resource, then these limits are summed, and that limit is
+assigned to the pod level cgroup.
+
+The implications of the current `Burstable` definition is that if any containers belonging to the Pod don't define a limit, then those containers
+are effectively not limited by the Linux kernel. Conversely (as happens in the `Guaranteed` QoS level), if all of the containers provided a limit, 
+then even though the pod level cgroup is configured with the sum of those limits, there is no significance to this since no container can ever use 
+more of the resource than what is defined in the container level cgroup.
+
+The proposal in this KEP is to allow users to opt-in to a slightly modified definition of the `Burstable` QoS level. In this modified definition,
+the sum of all defined container limits for each resource are always assigned to the pod-level cgroup. This is done by adding an attribute called 
+`shareBurstableLimits` on the Pod specification level. If this attribute is set to false (the default) the feature is disabled, and if it is set to 
```

Comment:
Compared to adding a bool value to accumulate the limit, why not consider adding a limit to the Pod. In a multi-container scenario, Pods are scheduled according to the sum of requests. We want to control the amount of burstable resources for the entire Pod to avoid sudden resource contention for the entire machine, such as oom.

Reply:
I've updated the KEP to move to setting the resources on the pod level instead of the `shareBurstableLimits` option.